[BEAM-8151, BEAM-7848] Swap to using an unbounded thread pool which shrinks when threads are not being used.

lukecwik opened the pull request:
I took a look at several threading libraries available to Python and none of them worked or had serious bugs. The best one was the collapsing-thread-pool-executor.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

lukecwik commented:
R: @pabloem @TheNeuralBit 
CC: @aaltay @robertwb @Hannah-Jiang

robertwb commented on sdks/python/apache_beam/runners/portability/fn_api_runner.py:
```diff
@@ -1139,7 +1139,7 @@ def __init__(self, state, provision_info, max_workers):
     self.provision_info = provision_info
     self.max_workers = max_workers
     self.control_server = grpc.server(
-        futures.ThreadPoolExecutor(max_workers=self.max_workers))
```
I wonder if we can get rid of `max_workers` altogether here (and below).

lukecwik replied:
I left it here since it is computed based upon the number of environments.

robertwb replied:
Yes, but we can simplify the code by omitting the computation altogether by making this (virtually) unbounded instead.

robertwb commented on sdks/python/apache_beam/runners/portability/local_job_service.py:
```diff
@@ -73,7 +73,7 @@ def __init__(self, staging_dir=None):
 
   def start_grpc_server(self, port=0):
-    self._server = grpc.server(futures.ThreadPoolExecutor(max_workers=3))
+    self._server = grpc.server(CollapsingThreadPoolExecutor(max_workers=3))
```
3 is arbitrary, perhaps just omit the argument?

robertwb commented on sdks/python/apache_beam/runners/portability/local_job_service.py:
```diff
@@ -183,7 +183,7 @@ def __init__(self, worker_command_line, control_address, worker_id=None):
 
   def run(self):
-    logging_server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
+    logging_server = grpc.server(CollapsingThreadPoolExecutor(max_workers=10))
```
Similarly, just use the default.

robertwb commented on sdks/python/apache_beam/runners/portability/portable_stager_test.py:
```diff
@@ -56,7 +56,7 @@ def _stage_files(self, files):
           name in staging location.
     """
-    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
+    server = grpc.server(CollapsingThreadPoolExecutor(max_workers=10))
```
Consider omitting the argument here (and for other tests, aside from artifact_serving_test where we want to control the parallelism).

lukecwik commented:
I removed it from everywhere except for the test that you mentioned and the one place where we have logic. Note that the default is 5 * cpu count (and not unlimited).

aaltay commented on sdks/python/setup.py:
```diff
@@ -104,6 +104,7 @@ def get_version():
 REQUIRED_PACKAGES = [
     'avro>=1.8.1,<2.0.0; python_version < "3.0"',
     'avro-python3>=1.8.1,<2.0.0; python_version >= "3.0"',
+    'collapsing-thread-pool-executor>="2018.6"',
```
Couple of questions about the dependency:
- Could we add an upper bound? (e.g. <2019)
- This does not look like a very used library with total 10 commits and no updates in the past year. Is there an alternative?

lukecwik replied:
We could add one but it would be arbitrary. WDYT?

I looked through Pypi and could not find another one, I looked at about 40 libraries that mentioned threadpool and this one was the best since it:
* mirrored the ThreadPoolExecutor API that is part of Python
* didn't have any additional dependencies
* I read the code and it was based upon the ThreadPoolExecutor implementation that is part of Python

aaltay replied:
Sounds good to me. Thank you for the detailed due diligence.

aaltay commented on sdks/python/container/base_image_requirements.txt:
```diff
@@ -25,6 +25,7 @@
 
 avro==1.8.2;python_version<="2.7"
 avro-python3==1.8.2;python_version>="3.4"
+collapsing-thread-pool-executor>="2018.6"
```
We pin versions in this file.

lukecwik replied:
I see, will revert

lukecwik replied:
Removed the change to this file.

aaltay replied:
It is fair to change this to 
`collapsing-thread-pool-executor=="2018.6"`

Removing is also fine.

aaltay commented on sdks/python/apache_beam/runners/worker/sdk_worker_main.py:
```diff
@@ -202,7 +202,7 @@ def _get_worker_count(pipeline_options):
                    experiment).group('worker_threads'))
 
-  return 12
+  return 10000
```
Do we know why 10K is a good number? Would not this practically accept all work?

How did we come up with 12 in the first place. (@angoenka may have information)

lukecwik replied:
I choose a value where it is effectively unbounded.

aaltay replied:
Do we have any benchmarks to look at and see what is the impact of this?

My underlying question is what does this imply for python streaming running with the java runner harness?

lukecwik replied:
It means that we will now schedule them to be executed while before we would block (which could lead to deadlock for specific pipeline shapes) where the execution of one bundle relies on the execution of another bundle.

aaltay replied:
OK. That actually sounds good, could fix some stuckness issues.

I am not very clear on whether this will impact performance or not. We can at least just watch the changes after the merge. (cc: @angoenka and @y1chi as an FYI.)

angoenka replied:
> Do we know why 10K is a good number? Would not this practically accept all work?
> 
> How did we come up with 12 in the first place. (@angoenka may have information)

It was just some basic experimentation where for the job which I was experimenting on did not show a lot of benefit beyond 12 threads.

For TFX pipelines, we recommend 100 threads when running on Flink because of some scheduling issues where we use to get dead lock in situation similar to the one mentioned by Luke.

I am not sure how it will impact the resource usage if we have 10000 standing threads for single core.

angoenka replied:
also +1 for this change if it does not have performance impact as it will solve a problem which we discussed over the mailing thread https://lists.apache.org/thread.html/0cbf73d696e0d3a5bb8e93618ac9d6bb81daecf2c9c8e11ee220c8ae@%3Cdev.beam.apache.org%3E

robertwb replied:
Yes, this is about avoiding deadlocks (there may be dependencies among the scheduled items that are not reflected in the order that they are received/processed) than performance.

lukecwik commented:
Run Python PreCommit

lukecwik commented:
Run Python PreCommit

pabloem commented:
Thanks Luke. This should be good for teh worker.

pabloem commented:
Run Python PreCommit

lukecwik commented:
Run Python PreCommit

lukecwik commented:
Run Python PreCommit

lukecwik commented:
Run Python PreCommit

lukecwik commented:
Run Python PreCommit

lukecwik commented:
Run Python PreCommit

aaltay commented:
Precommit is failing with:

16:09:15 collapsing-thread-pool-executor 2018.6 has requirement futures==3.2.0, but you have futures 3.3.0.
16:09:15 ERROR: InvocationError for command /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_Phrase/src/sdks/python/test-suites/tox/py2/build/srcs/sdks/python/target/.tox-py27-lint/py27-lint/bin/pip check (exited with code 1)
16:09:15 py27-lint run-test-post: commands[0] | /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_Phrase/src/sdks/python/test-suites/tox/py2/build/srcs/sdks/python/scripts/run_tox_cleanup.sh
16:09:15 ___________________________________ summary ____________________________________
16:09:15 ERROR:   py27-lint: commands failed

The reason is a dependency conflict.
collapsing-thread-pool-executor requires futures == 3.2.0 for python 2.7 (https://github.com/ftpsolutions/collapsing-thread-pool-executor/blob/master/setup.py#L22)

Beam's setup.py accepts "'futures>=3.2.0,<4.0.0; python_version < "3.0"',"

We can do a few things:
- Update collapsing-thread-pool-executor to relax the upper bound for futures dependency
- Update beam to restict its futures dependency to use 3.2.0

aaltay commented:
Proposed this change (https://github.com/ftpsolutions/collapsing-thread-pool-executor/pull/2) will if that will be accepted.

lukecwik commented:
Thanks Ahmet

TheNeuralBit commented:
Is https://github.com/ftpsolutions/collapsing-thread-pool-executor/issues/1 a concern?

lukecwik commented:
Ahmet's change was merged, waiting on release to PyPi

lukecwik commented:
Done, waiting on tests to pass and will merge.

lukecwik commented:
Run Python PreCommit

lukecwik commented:
Run Python PreCommit

pabloem commented:
Run Python PreCommit

lukecwik commented:
The issue is that the collapsing-thread-pool-executor has a bug where during shutdown it tries to join worker threads but the worker threads have a default 30 second timeout waiting for work.

lukecwik commented:
Started discussion looking for alternatives:
https://lists.apache.org/thread.html/03710e3d37cdbb7cd4711f9ffebf9741f5e88b6d595542246b2d14f2@%3Cdev.beam.apache.org%3E

robertwb commented:
Could we perhaps get this in guarded by a flag so we can do performance testing while we wait for fixes and another release?

lukecwik commented:
I think the complexity of guarding this with a flag will be much larger then rebasing when the library is updated once https://github.com/ftpsolutions/collapsing-thread-pool-executor/pull/4 is merged and released so I'm not seeing the value.

lukecwik commented:
Run Python PreCommit

lukecwik commented:
Run Python PreCommit

lukecwik commented:
Run Python PreCommit

lukecwik commented:
Run Python PreCommit

lukecwik commented:
Run Python PreCommit

lukecwik commented:
R: @robertwb @aaltay @angoenka 
This has now changed enough that it should be re-reviewed.

lukecwik commented:
Run Portable_Python PreCommit

lukecwik commented:
Run Python PreCommit

lukecwik commented:
I ran the performance benchmark using the fn_api_runner_microbenchmark:

before (at commit 80e8fb59b6d785eb462e6a05b91d029fd8ff50c6):
```
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 1 of 10, per element time cost: 3.0014 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 2 of 10, per element time cost: 0.0394958 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 3 of 10, per element time cost: 0.0220158 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 4 of 10, per element time cost: 0.0160263 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 5 of 10, per element time cost: 0.0130143 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 6 of 10, per element time cost: 0.0111442 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 7 of 10, per element time cost: 0.00963751 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 8 of 10, per element time cost: 0.00856183 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 9 of 10, per element time cost: 0.00796899 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 10 of 10, per element time cost: 0.00760294 sec


Fixed cost   3.49717886575063
Per-element  0.003790947336139101
R^2          0.963500601942131
```

after (including #10091 :
```
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 1 of 10, per element time cost: 2.8741 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 2 of 10, per element time cost: 0.0401369 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 3 of 10, per element time cost: 0.0219585 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 4 of 10, per element time cost: 0.0161115 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 5 of 10, per element time cost: 0.0124523 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 6 of 10, per element time cost: 0.0108133 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 7 of 10, per element time cost: 0.00963711 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 8 of 10, per element time cost: 0.00862622 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 9 of 10, per element time cost: 0.00795245 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 10 of 10, per element time cost: 0.00736677 sec


Fixed cost   3.4638128309249883
Per-element  0.003727303158153187
R^2          0.9529920831060255
```

which is a negligible change

aaltay commented on sdks/python/apache_beam/runners/worker/sdk_worker_main.py:
```diff
@@ -190,7 +190,7 @@ def _get_worker_count(pipeline_options):
 
   Returns:
-    an int containing the worker_threads to use. Default is 12
+    an int containing the worker_threads to use. Default is 10000.
```
Also update L206? (`return 12`)

lukecwik replied:
That was a mistake, in my follow-up #10091 I remove this method so I changed the comment to 12

aaltay commented on sdks/python/apache_beam/runners/worker/sdk_worker.py:
```diff
@@ -102,10 +102,9 @@ def __init__(
     #  Without wait, having multiple threads will not improve performance and
     #  will only add complexity.
-    self._progress_thread_pool = futures.ThreadPoolExecutor(max_workers=1)
+    self._progress_thread_pool = UnboundedThreadPoolExecutor()
```
Update the comment above? (`one thread is enough ...`)

lukecwik replied:
removed.

aaltay commented on sdks/python/apache_beam/utils/thread_pool_executor.py:
```diff
@@ -0,0 +1,163 @@
+
+
+class _WorkItem(object):
+  def __init__(self, f, fn, args, kwargs):
```
For readability could you rename `f` to `future`?

lukecwik commented:
Run Python PreCommit

lukecwik commented:
Run Portable_Python PreCommit

aaltay commented:
Run Python PreCommit
